packages/eslint-plugin/src/util/getThisExpression.ts

```diff
@@ -0,0 +1,26 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export function getThisExpression(
+  node: TSESTree.Node,
+): TSESTree.ThisExpression | undefined {
+  while (node) {
+    if (node.type === AST_NODE_TYPES.CallExpression) {
+      node = node.callee;
+    }
+
+    if (node.type === AST_NODE_TYPES.ThisExpression) {
+      return node;
+    }
+
+    if ('object' in node) {
+      node = node.object;
+    } else {
+      break;
+    }
```

Comment:
Just so we only ever handle one node per iteration - makes this more predictable

```suggestion
    if (node.type === AST_NODE_TYPES.CallExpression) {
      node = node.callee;
    } else if (node.type === AST_NODE_TYPES.ThisExpression) {
      return node;
    } else if ('object' in node) {
      node = node.object;
    } else {
      break;
    }
```